Warrior changes (latest)

parkscs opened the pull request:
I have continued working on this project since submitting the original pull request and I think it probably makes sense to close that first PR and create a new one to replace it based on the newest branch.  The warrior_changes_2 branch includes the recent work that I've done on this project.  We've had this code on the test server for some time now to collect feedback from the community and to identify bugs (of which we found and fixed a few), and at this point I think it probably makes sense to just skip warrior_changes_1 and move right to warrior_changes_2.

Because this includes the code from the original PR and I know some feedback was already given for that first PR, I've taken all the feedback from the first warrior_changes pull request and incorporated it into this branch (e.g., I removed the dead code @cizra flagged earlier today).  There's a lot here but I've done my best to keep the code separated into logical commits.

As an overview, the following PR includes the following updates to the warrior class:
- Updates to legacy warrior abilities (bodyslam, stomp, etc.) to make them worth using
- Refactoring of warrior abilities to move abilities into more logical disciplines and to spread the abilities around more evenly.  Also added 2h specialization to warriors.
- Refactoring deathstroke - this ability now deals more damage overall and scales inversely with victim health.  Added a lockout and a significant debuff to discourage tanks from using this offensive ability.
- New ability - Rally - short duration AoE buff that raises max hp/vitality
- New ability - Whirlwind - AoE attack with a lockout
- New ability - Focus Attack - a reliable additional autoattack with a lockout.
- New ability - Slam - a low damage melee attack with no lockout
- New ability - Fortify - drastically reduces incoming damage for the warrior for a very short duration.  Lockout and requires a shield.
- New ability - Bloodlust - combat and healing bonuses that stack as the warrior kills opponents.  Short duration and buffs will fall off if the warrior is out of combat for long.
- Berserk tweaks and new ability - advanced berserking - strengthens the base berserk ability and provides increased bloodlust stacks while berserking.

At this point, this is a fairly complete overhaul of the warrior class.  I anticipate there could be a few new additions in the future, but I think most changes will likely come in the form of minor buffs/nerfs/bug fixes as we get more testing/feedback.

Durhamite2118 commented:
To support this request, these changes were evaluated on the test server for ~30 days with no obvious bugs or balance issues in game play.  We are committed to the idea that if we discover an unrealized issue/bug/balance problem that we pull this back to reevaluate and correct as needed.  Thanks to Boz for his hard work.

drewhershey commented on code/code/cmd/cmd_rally.cc:
```diff
@@ -0,0 +1,83 @@
+  aff2.bitvector = 0;
+ 
+  // Loop for each person in room
+  int found = false;
+  for(StuffIter it=caster->roomp->stuff.begin(); it!=caster->roomp->stuff.end() && (t=*it); ++it) {
+    tmp_victim = dynamic_cast<TBeing *>(t);
+    if (!tmp_victim)
+      continue;
+    if (caster->inGroup(*tmp_victim)) {
+      act("You feel revitalized and ready to continue the battle!", TRUE, tmp_victim, NULL, NULL, TO_CHAR);
+
+      tmp_victim->affectJoin(caster, &aff1, AVG_DUR_NO, AVG_EFF_YES, false);
+      tmp_victim->affectJoin(caster, &aff2, AVG_DUR_NO, AVG_EFF_YES, false);
+      tmp_victim->addToMove(modifierValue);
+      tmp_victim->addToHit(modifierValue);
+
+      found = true;
+    }
+  }
+  if (!found)
+    caster->sendTo("But, there's nobody in your group.\n\r");  
```
```suggestion
  bool found = false;
  for (TThing *thing : roomp->stuff) {
    auto *person = dynamic_cast<TBeing *>(thing);

    if (!person || !inGroup(*person))
      continue;

    person->sendTo("You feel revitalized and ready to continue the battle!\n\r");
    person->affectJoin(this, &aff1, AVG_DUR_NO, AVG_EFF_YES, false);
    person->affectJoin(this, &aff2, AVG_DUR_NO, AVG_EFF_YES, false);
    person->addToMove(modifierValue);
    person->addToHit(modifierValue);

    if (person != this)
      found = true;
  }

  if (!found)
    sendTo("But, there's nobody in your group.\n\r");
```

drewhershey reviewed (changes requested):
Hopefully I wasn't too overbearing in my review! The raw number of comments makes it look like I found more problems than I actually did - I just tried to use lots of small, individual suggestions to make it easy to see what I was proposing be changed.

If I understand correctly how this works, for all of these suggestions except one you should be able to click either the "commit suggestion" (which will apply the suggestion as its own commit) or "add suggestion to batch" (to commit a group of suggestions together as a single commit) buttons in the individual comment to apply the suggestion straight to the code through Github. Pretty handy.

parkscs commented on code/code/cmd/cmd_bodyslam.cc:
```diff
@@ -221,50 +221,35 @@ static int bodyslamHit(TBeing *caster, TBeing *victim)
-    int modif = 1;
-    modif += (caster->getPrimaryHold() ? 0 : 2);
-    modif += (caster->getSecondaryHold() ? 0 : 1);
+  int bKnown = caster->getSkillValue(SKILL_HEADBUTT);
+  int successfulHit = caster->specialAttack(victim, SKILL_HEADBUTT);
+  int successfulSkill = caster->bSuccess(bKnown, SKILL_HEADBUTT);
```
That would make too much sense.  XD

parkscs commented:
Pulled in your changes Drew and rebased all the commits to break it back out.  A few things I had to tweak slightly (where you referenced the variable "weapon" I think was a mistake and I changed that to reference the variable name declared above for the primary hand weapon) but otherwise I think it's pretty much as suggested.  Any further thoughts/feedback are welcome.

drewhershey commented:
Everything works well in my testing after the recent commits 👍

parkscs reviewed (commented):
All looks reasonable to me.

cizra commented on code/code/cmd/cmd_deathstroke.cc:
```diff
@@ -5,222 +5,168 @@
   }
+
+  // Avoid players attacking un-harmable victims
   if (caster->noHarmCheck(victim))
```
No message sent? Or is that in noHarmCheck?

parkscs replied:
Updated to add a quick message.

drewhershey replied:
Just FYI, noHarmCheck already sends a message based on circumstances - that's why it didn't have one here initially.

https://github.com/sneezymud/sneezymud/blob/221a9c1abea6439a7f682b220cee8cdf904633c3/code/code/misc/offense.cc#L2066-L2112

cizra commented on code/code/cmd/cmd_deathstroke.cc:
```diff
@@ -5,222 +5,168 @@
                   FALSE, caster, 0, victim, TO_VICT);
-      if (caster->reconcileDamage(victim, 0,sktype) == -1)
-        return DELETE_VICT;
     }
```
else what? Can it happen? Print something?

parkscs replied:
I think the else is it just continues through the function without returning.  It goes on to allow the victim a chance to counter attack down below and ultimately will return an appropriate value, but there's not specifically an else case to handle here I believe.

cizra commented on code/code/cmd/cmd_deathstroke.cc:
Kinda weird to check for skills first and stun last.. but I guess it doesn't make any difference

parkscs replied:
Submitted an updated commit to re-order.

cizra reviewed (commented):
This branch saw very substantial changes since my last review. I need to reread.

cizra commented on code/code/disc/disc_brawling.cc:
```diff
@@ -29,9 +33,13 @@ CDBrawling::CDBrawling(const CDBrawling &a) :
+  skAdvBerserk(a.skAdvBerserk)
 {
 }
 
```
These copy constructors and assignment operators usually don't do anything new that the compiler wouldn't generate automatically. We should get rid of them all, tho maybe not in this  PR.

drewhershey replied:
I've tested removing all those constructors from about a dozen disc_x.cc files and letting them be replaced by implicit constructors and the code compiled fine, so I think you're right that they're not needed. One big PR getting rid of them all is on my to-do list.

parkscs replied:
Happy to help if I can with removing them/testing, but would prefer to do so in a new PR.

cizra commented on code/code/misc/limits.cc:
```diff
@@ -467,6 +467,22 @@ int TPerson::hitGain()
+    for (ch_affected = affected; ch_affected; ch_affected = ch_affected->next) {
+      if (ch_affected->type == SKILL_BLOODLUST) {
+        // set the mod and remove the affect so we can add it fresh
+	mod += ch_affected->modifier;
```
bad indent

parkscs replied:
I think I addressed the concern?  I had some tab characters in a couple indents, replaced with all space characters.  Let me know if the issue was something else.

cizra replied:
Yep, looks fine now.

cizra commented on lib/help/_skills/advanced berserking:
```diff
@@ -0,0 +1,5 @@
+a handful of Berserker tribes have truly mastered the art.  These advanced berserkers
+maintain all the reckless abandon of a typical berserker but with even further
+heightened combat prowess.  Moreover, particularly skilled berserkers can work themselves
+into a bloodlust the longer they fight.
```
There's some grammar missing here. They can work themselves into a _more intense_ bloodlust the longer they fight, something like that?

parkscs replied:
The grammar is correct, but I can reword to make it clearer.  The intent is to convey that bloodlust stacks can be gained in two ways - killing an enemy or, with advanced berserking, over time while fighting.

cizra commented on lib/help/_skills/bloodlust:
```diff
@@ -0,0 +1,17 @@
+the average adventurer.  However, this fervor of battle can fade as quickly
+as it comes on once a battle has ended.
+
+Rumor has it that some tribes of northern Berskerers have further mastered
```
Oooh, opportunity to tie this into racial subcategories!

parkscs replied:
We were floating around some ideas for synergies with various races and shapeshifts recently - definitely something to consider in the future, especially if it's not game-changing buffs/advantages but just adds a bit of flavor.

cizra commented on lib/help/_skills/whirlwind:
```diff
@@ -0,0 +1,9 @@
+
+Technical description: Whirlwind attempts to strike all non-grouped entities
+within the room.  A skill-success check is applied for attempting the maneuver,
+and then a separate hit-check is applied for each and every non-grouped entity.
```
Changes of this caliber merit an update of the news file

parkscs replied:
Pushed an updated news file - will need to update the date in the entry before this gets final approval/merged.
